### PR TITLE
Refactor registry.ImageGetter interface

### PR DESCRIPTION
**Pull request by albertogdd**
## Description

Make registry.ImageGetter stateful. Which removes a lot of unnecessary initializations of it.

## How can this be tested?

- Unit tests
- Deploy dynakube
- Deploy edgeconnect

```
make test/e2e/edgeconnect
make test/e2e/cloudnative/install
```

## Checklist

- [x] Unit tests have been updated/added
- [x] PR is labeled accordingly with a single label
- [x] I have read and understood the [contribution guidelines](https://github.com/Dynatrace/dynatrace-operator/blob/main/CONTRIBUTING.md)

### CODE REVIEW

File: src/controllers/dynakube/version/fakeregistry.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
+package version
```
Comment (albertogdd) [inline src/controllers/dynakube/version/fakeregistry.go:1]: Reminiscences from image_hash_test.go

File: src/controllers/dynakube/dynatraceclient/stub.go [hunk 2]
```diff
@@ -9,7 +9,7 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// StubBuilder can be used for unit tests where a full builder implementation is not needed
+// StubBuilder can be used for unit tests where a full builder.go implementation is not needed
```
Comment (andriisoldatenko) [inline src/controllers/dynakube/dynatraceclient/stub.go:12]:
```suggestion
// StubBuilder can be used for unit tests where a full builder implementation is not needed
```

File: src/controllers/dynakube/dynakube_controller.go [hunk 3]
```diff
@@ -216,6 +218,37 @@ func (controller *Controller) setupIstio(ctx context.Context, dynakube *dynatrac
 	return istioReconciler, nil
 }
 
+func (controller *Controller) createDynatraceClient(ctx context.Context, dynakube *dynatracev1beta1.DynaKube) (registry.ImageGetter, error) {
```
Comment (andriisoldatenko) [inline src/controllers/dynakube/dynakube_controller.go:221]:
```suggestion
func (controller *Controller) createDynatraceRegistryClient(ctx context.Context, dynakube *dynatracev1beta1.DynaKube) (registry.ImageGetter, error) {
```

**Review by 0sewa0 — CHANGES_REQUESTED** [review]

File: src/cmd/troubleshoot/image.go [hunk 4]
```diff
@@ -86,7 +86,23 @@ func tryImagePull(troubleshootCtx *troubleshootContext, image string) error {
...
+	}
+
+	var trustedCAs []byte
+	if troubleshootCtx.dynakube.Spec.TrustedCAs != "" {
+		trustedCAs, err = troubleshootCtx.dynakube.TrustedCAs(troubleshootCtx.context, troubleshootCtx.apiReader)
+		if err != nil {
+			return err
+		}
+	}
+	transport, err = registry.PrepareTransport(transport, proxy, trustedCAs)
```
[the remark this answers is not in the document]
Comment (andriisoldatenko) [reply src/cmd/troubleshoot/image.go:105]: I think it's a good idea, on it.

File: src/controllers/dynakube/dynakube_controller.go [hunk 5]
```diff
@@ -211,6 +214,28 @@ func (controller *Controller) setupIstio(ctx context.Context, dynakube *dynatrac
 	return istioReconciler, nil
 }
 
+func (controller *Controller) createDynatraceRegistryClient(ctx context.Context, dynakube *dynatracev1beta1.DynaKube) (registry.ImageGetter, error) {
+	keychain := dynakube.PullSecretWithoutData()
```
Comment (0sewa0) [inline src/controllers/dynakube/dynakube_controller.go:218]:
```suggestion
	pullSecret := dynakube.PullSecretWithoutData()
```
`keychain` is a bit confusing name, as we have `Keychain` as a specific part of the registry.
Also when saying `registry.WithKeyChainSecret(&keychain)` reads a bit backwards, like "the Secret for the Keychain is the Keychain it self". `registry.WithKeyChainSecret(&pullSecret)` read much nicer IMO

File: src/controllers/dynakube/dynakube_controller.go [hunk 6]
```diff
@@ -211,6 +214,28 @@ func (controller *Controller) setupIstio(ctx context.Context, dynakube *dynatrac
...
+	if err != nil {
+		return nil, err
+	}
+
+	registryClient, err := controller.registryClientBuilder(
+		registry.WithContext(ctx),
+		registry.WithApiReader(controller.apiReader),
+		registry.WithKeyChainSecret(&keychain),
+		registry.WithTransport(transport),
+	)
```
Comment (0sewa0) [inline src/controllers/dynakube/dynakube_controller.go:231]: 👨‍🍳's 💋

File: src/controllers/dynakube/version/updater.go [hunk 7]
```diff
@@ -96,12 +94,10 @@ func determineSource(updater versionStatusUpdater) status.VersionSource {
...
 }
 
-func setImageIDWithDigest( //nolint:revive
+func setImageIDWithDigest(
 	ctx context.Context,
-	apiReader client.Reader,
-	dynakube *dynatracev1beta1.DynaKube,
 	target *status.VersionStatus,
-	imageVersionFunc ImageVersionFunc,
+	registryClient registry.ImageGetter,
```
Comment (0sewa0) [inline src/controllers/dynakube/version/updater.go:100]: didn't think of that, but ⭐

**Review by 0sewa0 — APPROVED** [review]
Ran it with Dynakube(OA + AG) and EdgeConnect, practically LGTM.
Left 1 minor comment, if you agree address it and I will approve again 😉

File: src/controllers/dynakube/dynakube_controller.go (hunk 5 above)
Comment (andriisoldatenko) [reply src/controllers/dynakube/dynakube_controller.go]: fixed in c247cec7

**Review by andriisoldatenko — APPROVED** [review]

**Review by 0sewa0 — APPROVED** [review]

**Pull request merged**